src/main/java/org/jenkinsci/plugins/bitbucket/BitbucketBuildStatusNotifier.java

```diff
@@ -232,6 +229,57 @@ private BitbucketBuildStatus createBitbucketBuildStatusFromBuild(AbstractBuild b
         return new BitbucketBuildStatus(buildState, buildKey, buildUrl, buildName);
     }
 
+    private interface ScmAdapter {
+        URIish getRepositoryUri(SCM scm) throws Exception;
+        String getLastCommitId(AbstractBuild build) throws Exception;
+    }
+
+    private class GitScmAdapter implements ScmAdapter {
+
+        public URIish getRepositoryUri(SCM scm) throws Exception {
+            GitSCM gitSCM = (GitSCM) scm;
+            List<RemoteConfig> repoList = gitSCM.getRepositories();
+            if (repoList.size() != 1) {
+                throw new Exception("None or multiple repos");
+            }
+
+            return repoList.get(0).getURIs().get(0);
+        }
+
+        public String getLastCommitId(AbstractBuild build) throws Exception {
+            GitChangeSetList changeSets = (GitChangeSetList) build.getChangeSets().get(0);
+            List<GitChangeSet> gitChangeSet = changeSets.getLogs();
+            if (gitChangeSet.size() < 1) {
+                throw new Exception("No commits found on change set to send to Bitbucket!");
+            }
+
+            return gitChangeSet.get(gitChangeSet.size() - 1).getCommitId();
+        }
+    }
+
+    private class MercurialScmAdapter implements ScmAdapter {
+
+        public URIish getRepositoryUri(SCM scm) throws Exception {
+            MercurialSCM hgSCM = (MercurialSCM) scm;
+
+            String source = hgSCM.getSource();
+            if (source == null || source.isEmpty()) {
+                throw new Exception("None or multiple repos");
+            }
+
+            return new URIish(source);
+        }
+
+        public String getLastCommitId(AbstractBuild build) throws Exception {
+            MercurialChangeSetList changeSets = (MercurialChangeSetList) build.getChangeSets().get(0);
+            List<MercurialChangeSet> hgChangeSet = changeSets.getLogs();
+            if (hgChangeSet.size() < 1) {
+                throw new Exception("No commits found on change set to send to Bitbucket!");
+            }
+            return hgChangeSet.get(hgChangeSet.size() - 1).getRevision();
```

Comment:
Here is the line to fix, somehow the change set of mercurial are inverse ordered, so index 0 is last commit and size-1 is the first commit. I don't know if this could be a bug on the mercurial plugin but anyways we should fix it in this plugin.